db/commitlog/commitlog.cc

```diff
@@ -1494,6 +1491,45 @@ future<> db::commitlog::segment_manager::do_writes(future<> next, std::vector<me
     return next;
 }
 
+future<> db::commitlog::segment_manager::do_shutdown(future<> next, std::vector<message>& messages) {
+    bool shutdown = !message_partition(messages, shutdown_request_index).empty();
+    if (shutdown) {
+        // Wait for all pending requests to finish. Need to sync first because segments that are
+        // alive may be holding semaphore permits.
+        auto block_new_requests = get_units(_request_controller, max_request_controller_units());
+        try {
+            clogger.debug("Begin shutdown sequence");
+
+            co_await sync_all_segments();
+            co_await std::move(block_new_requests);
+
+            _state = state::shutting_down; // please do not create new segments etc.
+
+            discard_unused_segments();
+            co_await do_pending_deletes();
+
+            auto ep = std::make_exception_ptr(shutdown_marker{});
+            if (_recycled_segments.empty()) {
+                abort_recycled_list(ep);
+            }
+            abort_deletion_promise(ep);
+
+            co_await shutdown_all_segments();
+            co_await clear_reserve_segments();
+
+            _state = state::terminated;
+
+            clogger.debug("Shutdown sequence complete");
+
+            wakeup(); // ensure we exit message loop now
+        } catch (...) {
+            clogger.error("Exception during shutdown: {}. Aborting.", std::current_exception());
+            abort();
+        }
+    }
+    co_await std::move(next);
```

Comment:
Why can't the caller co_await next, instead of sending it here for us to co_await it?

I'd like to see more boring code. The more we can make it resemble sequential code (with concurrency only where truly needed) the easier it will be to understand.

Reply:
Which is why it looks like this. We _do_ want to maintain the continuation chain (so we have a waitable achor). But shutdown is to many lines (imho) to make is a lambda-wrapper like the write/flush calls. So we co_await. 
Again, the other option is to keep every task in a giant (chunked) vector, but then we have to reap and gc it -> more complex imho. 
I can add some abstraction here "add_task(Func&&)" that simply wraps a -> future<> call so we never see them in the actual handler routines, but that would essentially add another coroutine frame to every task chain link (or I write it using raw futures -> extra malloc anyway). This is basically trying to keep the spurious frames down.

Reply:
The idiom

    _done = _done.then([] { blah });

is well known an accepted. It's risky in that it doesn't bound concurrency, but that can be explained in the comment.

What I don't understand is

   co_await foo(std::move(some_future));
   ...

   foo(future<> next) {
         ...
         co_await next;
   }

Why can't the caller co_await some_future after foo() returns?

Reply:
> 
> 
> The idiom
> 
> ```
> _done = _done.then([] { blah });
> ```
> 
> is well known an accepted. It's risky in that it doesn't bound concurrency, but that can be explained in the comment.
> 
> What I don't understand is
> 
> co_await foo(std::move(some_future));
> ...
> 
> foo(future<> next) {
> ...
> co_await next;
> }
> 
> Why can't the caller co_await some_future after foo() returns?

Those are not the same. 
`_done = _done.then([] { <something>})`
executes `<something>` _after current `_done` task. Including initiating the operations in _done.
To ensure we actually do them in parallel we do
`_done = [f = std::move(_done)] mutable { return when_all(<something>, std::move(f)); };`
But expressed coroutine-ish to avoid extra allocs etc- 

The point being that we basically want a mutable `when_all` set that we can incrementally add to.

Reply:
I meant "initiating the operations in <something>".

Reply:
And yes, this could be expressed with the when_all for coroutines you added to seastar, but it would require the somewhat ugly hide-actual-future-in-lambda and would cause an additional coroutine frame.